### ISSUE TITLE
Training head rotation classifier

### ISSUE THREAD

**Issue by leg0m4n** [phizaz/diffae#8]
Hi Konpat and team of DiffAE,

I m still trying to train head rotation classifier, using data from FFHQ-Aging(they labeled the headpose).
I am using:
* 1 layer linear regression from cond to headpose.(512,3)
* MSEloss with a learning rate of 0.001
* Training data: 1200 images and headposes.
* batch size of 128(tried everything from 16 to 512)
* normalized cond, but not the headpose output. Would you recommend normalizing output as well?
* accumulate_grad_batches=7
* gradient_clip_val=0.5
* 300 epoches

For some reason, loss is keep being stuck around 65, which seems to be too large. Do you have any idea, what I am doing wrong and how I can improve this loss? Would you recommend using more data(perhaps all 70k images)? In your examples, 1k labeled images were enough, so I m not sure increasing the size of training data will help.

Many thanks,
Rich

**Comment by leg0m4n** [comment]
I did a silly mistake of putting Relu after linear layer, which I guess is unnecessary, since its regression, after removing it the loss dropped to 20. Still doesn't seem like a good result. Adding one more layer(now its (512, 256) -> (256, 3), where 3 is number of classes - pitch, yaw, roll) helped to drop loss to around 14. Still doesn't seem like a good result. Very curious to learn from you h

**Comment by leg0m4n** [comment]
Sorry, the key slipped. Anyway, looking forward to your advice. Thank you!

**Comment by phizaz** [comment]
In my opinion, the head pose target (or any target) should be normalized to keep the output's range in check. Since your loss is MSE, the size of the loss depends so much on the "range" of the target. You cannot look only at the "loss" and decide whether it's too high or low. You should need a better metric.

**Comment by leg0m4n** [comment]
Ohhh, right, of course, targets should be normalized as well!

What other metric would you recommend using as a secondary marker to check the training performance?

**Comment by phizaz** [comment]
Not an expert here. One possible way for regression is R^2 (R-squared).

**Comment by leg0m4n** [comment]
Tried using the R2, indeed using both gives a better picture. Thank you for the tip! The current model gives around 80% at R2. Not great, not terrible. I'll try to train on the whole dataset with labels(70k) later on, I m pretty sure I can push it above 90%.

Meanwhile, I am trying to change cond based on this head rotation net. In your manipulate.ipynb you have this line
> cond2 = cond2 + 0.3 * math.sqrt(512) * F.normalize(cls_model.classifier.weight[cls_id][None, :], dim=1)

In my case, I just have a regression model (512) -> (3), do you know, by chance, how can I get the inverse of the direction of rotation(just one of three-axis) and get this second term(cls_model.classifier.weight[cls_id][None, :], dim=1)? E.g. I pass cond to my model, get the pitch, yaw, roll, and want to move the cond just along the pitch axis.

**Comment by phizaz** [comment]
Say your regression model is a linear regression model with a weight matrix (512, 3). The head rotation vector is one of the columns (512, 1) of the (512, 3) matrix. If your model is non-linear, e.g. MLP, the easiest way to get the head rotation vector is through grad(rotation prediction) wrt. latent code. Note that you'll get only a local direction in the case of the non-linear predictor.

> I'll try to train on the whole dataset with labels(70k) later on, I m pretty sure I can push it above 90%.

Here, I'm not sure we are on the same page. Training the autoencoder (DiffAE) does not require any label. I'm not sure, then, how would you utilize the labels?

**Comment by leg0m4n** [comment]
Oh, I see, so just the grad of the forward propagation, makes sense.

My current idea is to build a more precise model to operate in the latent space of cond(not training the whole diffae, just a more advanced classifier). In your paper classifiers are pretty binary(e.g. blond/not blond hair), and since my final goal is to control the motion of the faces(lips movement, visemes, based on speech audio), I decided to try something easier first and to rotate the head. For a precise head pose movement, the labels are needed to link it to the real parameters(degrees of rotation)

**Comment by YAwei666** [comment]
> Oh, I see, so just the grad of the forward propagation, makes sense.
> 
> My current idea is to build a more precise model to operate in the latent space of cond(not training the whole diffae, just a more advanced classifier). In your paper classifiers are pretty binary(e.g. blond/not blond hair), and since my final goal is to control the motion of the faces(lips movement, visemes, based on speech audio), I decided to try something easier first and to rotate the head. For a precise head pose movement, the labels are needed to link it to the real parameters(degrees of rotation)

hi, I also try to training a regression model, did you succeed in your head pose movement? and how you do this formula cond2 = cond2 + 0.3 * math.sqrt(512) * F.normalize(cls_model.classifier.weight[cls_id][None, :], dim=1)?
